release/pkg/assets/assets.go

```diff
@@ -0,0 +1,122 @@
+package assets
+
+import (
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	assetconfig "github.com/aws/eks-anywhere/release/pkg/assets/config"
+	assettypes "github.com/aws/eks-anywhere/release/pkg/assets/types"
+	"github.com/aws/eks-anywhere/release/pkg/filereader"
+	"github.com/aws/eks-anywhere/release/pkg/git"
+	releasetypes "github.com/aws/eks-anywhere/release/pkg/types"
+	sliceutils "github.com/aws/eks-anywhere/release/pkg/util/slices"
+)
+
+func getAssetsFromConfig(ac *assettypes.AssetConfig, rc *releasetypes.ReleaseConfig, eksDReleaseChannel, eksDReleaseNumber, kubeVersion string) ([]releasetypes.Artifact, error) {
+	var gitTag string
+	var err error
+	var artifacts []releasetypes.Artifact
+	var imageTagOverrides []releasetypes.ImageTagOverride
+	projectName := ac.ProjectName
+	projectPath := ac.ProjectPath
+	sourcedFromBranch := rc.BuildRepoBranchName
+	gitTagPath := projectPath
+	if ac.HasSeparateTagPerReleaseBranch {
+		gitTagPath = filepath.Join(projectPath, eksDReleaseChannel)
+	}
+	// Get git tag for project if exists
+	if !ac.NoGitTag {
+		if ac.AssignEksATag {
+			if rc.DevRelease {
+				tagList, err := git.GetRepoTagsDescending(rc.CliRepoSource)
+				if err != nil {
+					return nil, errors.Cause(err)
+				}
+				gitTag = strings.Split(tagList, "\n")[0]
+			} else {
+				gitTag = rc.ReleaseVersion
+			}
+		} else {
+			gitTag, err = filereader.ReadGitTag(gitTagPath, rc.BuildRepoSource, rc.BuildRepoBranchName)
+			if err != nil {
+				return nil, errors.Cause(err)
+			}
+		}
+	} else {
+		gitTag = "non-existent"
+	}
```

Comment:
This is what I was suggetsing at the beggining to try to implement options through closures/interfaces instead of boolean. The complexity of this block of code is significant, up to 4 nested ifs.

And there is very little we can do about it with boolean options: the more options we add, the more permutations.

Reply:
Fixed this.